packages/docusaurus-theme-common/src/internal.ts

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+// This re-export permits to handle some level of retro-compatibility
+// Users might swizzle unsafe components, using our internal apis.
+// When we move an API from internal to public, former internal imports
+// should keep working => less annoying for users.
+// TODO: how do we detect legacy internal imports in our own code?
+ 
+// export * from './index';
+
+export {
+  type Navbar,
+  type NavbarItem,
+  type NavbarLogo,
+} from './utils/useThemeConfig';
+export {
+  DocSidebarItemsExpandedStateProvider,
+  useDocSidebarItemsExpandedState,
+} from './contexts/docSidebarItemsExpandedState';
+export {DocsVersionProvider, useDocsVersion} from './contexts/docsVersion';
+export {DocsSidebarProvider, useDocsSidebar} from './contexts/docsSidebar';
+export {DocProvider, useDoc} from './contexts/doc';
+export type {DocContextValue} from './contexts/doc';
+
+export {createStorageSlot, listStorageKeys} from './utils/storageUtils';
```

Comment:
This api is only used in theme-common: users can't "see" it in practice after swizzling, so I don't think it's useful to commit to retrocompatibility on something like that. The only place where it's imported from entrypoint is `registerSw.ts` of the pwa plugin, which can't be swizzled anyway.

I'd rather keep this internal for now

Reply:
It's very useful for external theme authors. As explained in my comment above ^ "Public" also means site creators and theme authors are encouraged to use these abstractions, which creates a uniform UX. We use it on our own website, for example.

Reply:
ok to make it public 👍

But do we really want to document those and encourage theme authors and site owners to use those? 

What is the advantage for them compared to directly using `localStorage`?

We built this because we want configurable storage + failsafe for some very fancy edge cases, otherwise using browser storage APIs may be good enough?

Reply:
Yes, failsafe sounds like a compelling reason (and uniform log messages in that case). The abstraction is also quite good.